lib/src/main/kotlin/com/swmansion/starknet/data/serializers/TransactionSerializer.kt

```diff
@@ -0,0 +1,90 @@
+package com.swmansion.starknet.data.serializers
+
+import com.swmansion.starknet.data.types.*
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.descriptors.PrimitiveKind
+import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.json.*
+
+internal val transactionIgnoredKeys = listOf("transaction_hash", "contract_address")
+
+internal object TransactionSerializer : KSerializer<Transaction> {
+    override val descriptor = PrimitiveSerialDescriptor("Transaction", PrimitiveKind.STRING)
+
+    override fun deserialize(decoder: Decoder): Transaction {
+        require(decoder is JsonDecoder)
+        val element = decoder.decodeJsonElement()
+        val jsonElement = element.jsonObject
+        val typeElement = jsonElement.getOrElse("type") { throw SerializationException("Input element does not contain mandatory field 'type'") }
+
+        val type = decoder.json.decodeFromJsonElement(TransactionType.serializer(), typeElement)
+
+        return when (type) {
+            TransactionType.INVOKE -> deserializeInvoke(decoder, element)
+            TransactionType.DECLARE -> deserializeDeclare(decoder, element)
+            TransactionType.DEPLOY_ACCOUNT -> deserializeDeployAccount(decoder, element)
+            TransactionType.DEPLOY -> Json.decodeFromJsonElement(DeployTransaction.serializer(), element)
+            TransactionType.L1_HANDLER -> Json.decodeFromJsonElement(L1HandlerTransaction.serializer(), element)
+        }
+    }
+
+    override fun serialize(encoder: Encoder, value: Transaction) {
+        require(encoder is JsonEncoder)
+        require(value is ExecutableTransaction) {
+            "TransactionSerializer can only serialize ExecutableTransaction instances."
+        }
+        val jsonObject = when (value) {
+            is InvokeTransactionV3 -> encoder.json.encodeToJsonElement(ExecutableTransactionSerializer, value).jsonObject
+            is InvokeTransactionV1 -> encoder.json.encodeToJsonElement(ExecutableTransactionSerializer, value).jsonObject
+            is DeclareTransactionV3 -> encoder.json.encodeToJsonElement(ExecutableTransactionSerializer, value).jsonObject
+            is DeclareTransactionV2 -> encoder.json.encodeToJsonElement(ExecutableTransactionSerializer, value).jsonObject
+            is DeclareTransactionV1 -> encoder.json.encodeToJsonElement(ExecutableTransactionSerializer, value).jsonObject
+            is DeployAccountTransactionV3 -> encoder.json.encodeToJsonElement(ExecutableTransactionSerializer, value).jsonObject
+            is DeployAccountTransactionV1 -> encoder.json.encodeToJsonElement(ExecutableTransactionSerializer, value).jsonObject
```

Comment:
this looks like it could maybe be refactored?

Reply:
What is an overhead for this refactor? Have you considered getting rid of when altogether?
I *think* you could just define a sealed class and make it a parent for all those classes. Then you could have just something like:

```
val jsonObject = encoder.json.encodeToJsonElement(ExecutableTransactionSerializer, value).jsonObject
```

where ExecutableTransaction is your parent class.

Reply:
:eyes: oops, correct - we can simplify it this way.